### ISSUE TITLE
Show example of using terminal hash parameter?

### ISSUE THREAD

**Issue by gaand** [ga-wdi-boston/ruby-vs-js-hash-dictionary#9]
This is how rails handles method config options.

I haven't found many examples in plain ruby, other than [g]sub:

```ruby
'bar baz'.gsub /b[^[:space:]]+/, 'bar' => 'snafu', 'baz' => 'fubar'
# => "snafu fubar"
```

The thing to note is that the terminal hash parameter isn't surrounded by curly braces.

**Comment by neugierige** [comment]
@gaand I think that this would create more confusion than it's worth. However, if you feel strongly about it and want this to be added for WDI15, please let me know.

**Comment by gaand** [comment]
@neugierige Let's leave it on hold for now.

**Comment by payne-chris-r** [comment]
I completely agree with @gaand on this one. It's good to show. Basically the way I show this is by showing an example and than "desugaring" it:

``` ruby
'bar baz'.gsub /b[^[:space:]]+/, 'bar' => 'snafu', 'baz' => 'fubar'
# => "snafu fubar"
```

IS THE SAME AS:

``` ruby
'bar baz'.gsub({ /b[^[:space:]]+/, 'bar' => 'snafu', 'baz' => 'fubar'})
# => "snafu fubar"
```

**Comment by payne-chris-r** [comment]
I actually think it desugars to this...

``` ruby
'bar baz'.gsub( /b[^[:space:]]+/, {'bar' => 'snafu', 'baz' => 'fubar'})
# => "snafu fubar"
```

**Comment by neugierige** [comment]
@payne-chris-r Ok. When you show this, what example do you give to demonstrate it's usefulness?

**Comment by payne-chris-r** [comment]
I think @gaand's is ok, though I agree that it might be more confusing than it needs to be. I found something on stack overflow here: http://stackoverflow.com/questions/16576477/how-do-methods-use-hash-arguments-in-ruby

What about:

``` ruby
foo "regular argument", a: 12, :b => 13
```

where {a: 12} is syntactic sugar for {:a => 12}

and the entire line de-sugars to:

``` ruby
foo("regular argument", { :a => 12, :b => 13 })
```

**Comment by payne-chris-r** [comment]
That last example is trying to show BOTH the two different ways of using symbols as keys in `ruby` and ISN'T focusing solely on hash options, so maybe just do one of the following as starter code to focus in more on the point at hand:

``` ruby
foo "regular argument", a: 12, b: 13
```

OR

``` ruby
foo "regular argument", :a => 12, :b => 13
```

**Comment by neugierige** [comment]
got it, thanks for the clarification and the link @payne-chris-r 

@gaand, I'll default to you on this. I am concerned that this will open a can of worm, and I want to err on the side of caution because we are squishing this talk and the ruby-arrays into one afternoon.

**Comment by gaand** [comment]
@payne-chris-r, @neugierige  This issue should wait for the next iteration.

**Comment by MicFin** [comment]
@gaand can you verify that this issue should be implemented in the README and how/where it should be introduced?  

My addition to the thread is that we are introducing some "syntactic sugar" of methods when we introduce code like `foo("regular arg", a: 12, :b => 13)`.  

Will this add confusion since they have not been introduced to the longer form syntax `foo("regular arg", { a: 12, :b => 13 })` yet?  If we add the shorthand, should the longer syntax be shown alongside it to further illustrate what is going on?

The introduction to methods at this point will have been from https://github.com/ga-wdi-boston/ruby#ruby--methods and studies.  

Just let me know how you all saw this fitting into the lesson and I will adjust accordingly.

**Comment by jrhorn424** [comment]
> Will this add confusion since they have not been introduced to the longer form syntax foo("regular arg", { a: 12, :b => 13 }) yet

IMO, no.

> If we add the shorthand, should the longer syntax be shown alongside it to further illustrate what is going on?

Yes. FWIW, when I deliver ruby talks I always de-sugar as soon as there are questions about syntax.

We have some loose conventions around introducing ruby, but one I feel strongly about is preferring `{foo: 'bar'}` to `{:foo => 'bar'}`.

Additionally, I always use parens when invoking a method with arguments because it saves my ass once in a while by avoiding a parser error, I think it looks better, and it doesn't take long to type. So I prefer `foo('bar')` to `foo 'bar'`. The notable exception is when using macros, like `attr_reader` and friends, and rails-y things like `belongs_to :foo`. I feel less strongly about this, but I do believe we should be consistent across talks.

**Comment by gaand** [comment]
I also try to illustrate conventional ruby using rubocop's complaints.
